File: app/exports/jsonld_snippet.py
"""JSON-LD HTML snippet (Stage 4c). Script-safe: '<' → \\u003c so </script> cannot break out."""

from __future__ import annotations

import json
from typing import Any, Sequence

from app.exports.schema_org import build_organization_jsonld


def script_safe_json(data: Any) -> str:
    text = json.dumps(data, ensure_ascii=False, separators=(",", ":"))
    return (
        text.replace("<", "\\u003c")
        .replace("\u2028", "\\u2028")
        .replace("\u2029", "\\u2029")
    )


def organization_jsonld_snippet(site: Any, entities: Sequence[Any]) -> str:
    payload = script_safe_json(build_organization_jsonld(site, entities))
    return f'<script type="application/ld+json">{payload}</script>\n'

File: php/src/Exports/JsonLdSnippet.php
<?php
declare(strict_types=1);

namespace Bkbs\Exports;

/** Stage 4c: script-safe JSON-LD HTML snippet ('<' → \u003c). */
final class JsonLdSnippet
{
    public static function scriptSafeJson(mixed $data): string
    {
        $json = json_encode($data, JSON_UNESCAPED_UNICODE | JSON_UNESCAPED_SLASHES);
        if (!is_string($json)) {
            $json = '{}';
        }
        return str_replace(
            ["<", "\u{2028}", "\u{2029}"],
            ['\u003c', '\u2028', '\u2029'],
            $json
        );
    }

    /**
     * @param array<string, mixed> $site
     * @param list<array<string, mixed>> $entities
     */
    public static function organization(array $site, array $entities): string
    {
        $payload = self::scriptSafeJson(SchemaOrg::organization($site, $entities));
        return '<script type="application/ld+json">' . $payload . "</script>\n";
    }
}
